File: Array/Medium/NextPermutation.py
class Solution:
    def nextPermutation(self, nums):
        # Find the breakpoint of the array
        break_index = -1
        for i in range(len(nums)-2, -1, -1):
            if nums[i] < nums[i+1]:
                break_index = i
                break
        print(break_index)
        if break_index == -1:
            nums.reverse()  # Reverse the array in place
            #Reverse this array in place
            return 

        # Second Step => Get the closest greater element than the breakpoint number
        for i in range(len(nums)-1, break_index, -1):
            if nums[i] > nums[break_index]:
                # Swap the breakpoint element with the closest greater element
                nums[i], nums[break_index] = nums[break_index], nums[i]
                break

        # Third Step => Reverse the sub-array after the breakpoint to get the next permutation
        nums[break_index+1:] = reversed(nums[break_index+1:])

File: Array/NextPermutation.java
public class NextPermutation {

    public void nextPermutation(int[] nums) {
        // Find the index of the breakpoint in the array
        int breakIndex = -1;
        for (int i = nums.length - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                breakIndex = i;
                break;
            }
        }
        System.out.println(breakIndex);

        // If no breakpoint was found, reverse the entire array in place
        if (breakIndex == -1) {
            int start = 0, end = nums.length - 1;
            while (start < end) {
                int temp = nums[start];
                nums[start] = nums[end];
                nums[end] = temp;
                start++;
                end--;
            }
            return;
        }

        // Find the closest greater element than the element at the breakpoint
        for (int i = nums.length - 1; i > breakIndex; i--) {
            if (nums[i] > nums[breakIndex]) {
                // Swap the element at the breakpoint with the closest greater element
                int temp = nums[i];
                nums[i] = nums[breakIndex];
                nums[breakIndex] = temp;
                break;
            }
        }

        // Reverse the sub-array after the breakpoint to get the next permutation
        int start = breakIndex + 1, end = nums.length - 1;
        while (start < end) {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
}